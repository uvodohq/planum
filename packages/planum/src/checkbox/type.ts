import type {
  AriaCheckboxGroupItemProps,
  AriaCheckboxGroupProps,
  AriaCheckboxProps,
  ToggleProps,
} from '@react-types/checkbox'
import type { ReactNode } from 'react'

import type { CSS } from '../theme'
import type { StyledCheckboxVariants } from './checkbox.styles'

export interface CustomProps {
  css?: CSS
  label?: string
  children?: string
}

export type CheckboxProps = CustomProps &
  StyledCheckboxVariants &
  ToggleProps &
  AriaCheckboxProps

export type CheckboxGroupItemProps = CheckboxProps & AriaCheckboxGroupItemProps

export interface CheckboxBaseProps {
  isDisabled?: boolean
  isIndeterminate?: boolean
  isSelected: boolean
  isFocusVisible?: boolean
  label?: any
  pressProps?: any
  inputProps?: any
  focusProps?: any
  children?: any
  inputRef?: any
}

export type CheckboxGroupProps = AriaCheckboxGroupProps & {
  children: ReactNode
  css?: CSS
}
