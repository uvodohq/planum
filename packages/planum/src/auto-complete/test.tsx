import {
  autoUpdate,
  size,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useListNavigation,
  useRole,
} from '@floating-ui/react-dom-interactions'
import React, { forwardRef, useLayoutEffect, useRef, useState } from 'react'

interface ItemProps {
  children: React.ReactNode
  active: boolean
}

const Item = forwardRef<
  HTMLDivElement,
  ItemProps & React.HTMLProps<HTMLDivElement>
>(({ children, active, ...rest }, ref) => {
  const id = useId()
  return (
    <div
      ref={ref}
      role="option"
      id={id}
      aria-selected={active}
      {...rest}
      style={{
        background: active ? 'lightblue' : 'none',
        padding: 4,
        cursor: 'default',
        ...rest.style,
      }}>
      {children}
    </div>
  )
})
Item.displayName = 'Item'

export function AutoCompleteTest({ data }) {
  const [open, setOpen] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [activeIndex, setActiveIndex] = useState<number | null>(null)

  const listRef = useRef<Array<HTMLElement | null>>([])

  const { x, y, reference, floating, strategy, context, refs } =
    useFloating<HTMLInputElement>({
      whileElementsMounted: autoUpdate,
      open,
      onOpenChange: setOpen,
      middleware: [
        size({
          apply({ rects, elements }) {
            Object.assign(elements.floating.style, {
              width: `${rects.reference.width}px`,
              maxHeight: `${200}px`,
            })
          },
          padding: 10,
        }),
      ],
    })

  useLayoutEffect(() => {
    // IMPORTANT: When the floating element first opens, this runs when the
    // styles have **not yet** been applied to the element. This can cause an
    // infinite loop as `size` has not yet limited the maxHeight, so the whole
    // page tries to scroll. We must wrap it in rAF.
    requestAnimationFrame(() => {
      if (activeIndex != null) {
        listRef.current[activeIndex]?.scrollIntoView({ block: 'nearest' })
      }
    })
  }, [activeIndex])

  const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions(
    [
      useRole(context, { role: 'listbox' }),
      useDismiss(context),
      useListNavigation(context, {
        listRef,
        activeIndex,
        onNavigate: setActiveIndex,
        virtual: true,
        loop: true,
        focusItemOnHover: false,
      }),
    ],
  )

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const value = event.target.value
    setInputValue(value)

    if (value) {
      setOpen(true)
      setActiveIndex(0)
    } else {
      setOpen(false)
    }
  }

  const items = data.filter((item) =>
    item.toLowerCase().startsWith(inputValue.toLowerCase()),
  )

  return (
    <>
      <input
        {...getReferenceProps({
          ref: reference,
          onChange,
          value: inputValue,
          placeholder: 'Enter fruit',
          'aria-autocomplete': 'list',
          onKeyDown(event) {
            if (
              event.key === 'Enter' &&
              activeIndex != null &&
              items[activeIndex]
            ) {
              setInputValue(items[activeIndex])
              setActiveIndex(null)
              setOpen(false)
            }

            if (event.key === 'Tab') {
              setOpen(false)
            }
          },
        })}
      />
      {open && (
        <div
          {...getFloatingProps({
            ref: floating,
            style: {
              position: strategy,
              left: x ?? 0,
              top: y ?? 0,
              background: '#eee',
              color: 'black',
              overflowY: 'auto',
            },
          })}>
          {items.map((item, index) => (
            // eslint-disable-next-line react/jsx-key
            <Item
              {...getItemProps({
                key: item,
                ref(node) {
                  listRef.current[index] = node
                },
                onClick() {
                  setInputValue(item)
                  setOpen(false)
                  refs.reference.current?.focus()
                },
              })}
              active={activeIndex === index}>
              {item}
            </Item>
          ))}
        </div>
      )}
    </>
  )
}
