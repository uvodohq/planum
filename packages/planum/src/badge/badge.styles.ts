import type * as Stitches from '@stitches/react'

import { paragraphCss } from '../text'
import { styled } from '../theme'

export const StyledBadge = styled('span', paragraphCss, {
  px: '$8',
  borderRadius: '$pill',
  border: '1px solid transparent',
  color: '$textDark',
  display: 'inline-flex',
  alignItems: 'center',
  height: '$24',
  whiteSpace: 'nowrap',

  variants: {
    variant: {
      // filled variants
      warning: {
        bg: '$warning200',
      },
      neutral: {
        bg: '$surface200',
      },
      success: {
        bg: '$success200',
      },
      danger: {
        bg: '$danger200',
      },
      secondary: {
        bg: '$secondary',
      },
      primary: {
        bg: '$primary500',
        color: '$white',
      },

      // outline variants
      outlineWarning: {
        color: '$warning600',
        borderColor: '$warning600',
      },
      outlineNeutral: {
        color: '$neutral600',
        borderColor: '$neutral600',
      },
      outlineSuccess: {
        color: '$success600',
        borderColor: '$success600',
      },
      outlineDanger: {
        color: '$danger600',
        borderColor: '$danger600',
      },
      outlineSecondary: {
        color: '$secondary',
        borderColor: '$secondary',
      },
      outlinePrimary: {
        color: '$primary600',
        borderColor: '$primary600',
      },
    },
    bordered: {
      true: {},
    },
    uppercase: {
      true: {
        textTransform: 'uppercase',
      },
    },

    roundness: {
      sm: {
        borderRadius: '$sm',
      },
      pill: {
        borderRadius: '$pill',
      },
    },
  },
  defaultVariants: {
    variant: 'neutral',
    roundness: 'pill',
  },
  compoundVariants: [
    {
      bordered: true,
      variant: 'outlinePrimary',
      css: {
        borderColor: '$primary600',
      },
    },
    {
      bordered: true,
      variant: 'warning',
      css: {
        borderColor: '$warning300',
      },
    },
    {
      bordered: true,
      variant: 'neutral',
      css: {
        borderColor: '$surface300',
      },
    },
    {
      bordered: true,
      variant: 'warning',
      css: {
        borderColor: '$warning300',
      },
    },
    {
      bordered: true,
      variant: 'success',
      css: {
        borderColor: '$success300',
      },
    },
    {
      bordered: true,
      variant: 'danger',
      css: {
        borderColor: '$danger300',
      },
    },
    {
      bordered: true,
      variant: 'secondary',
      css: {
        borderColor: '$secondary100',
      },
    },
    {
      bordered: true,
      variant: 'primary',
      css: {
        borderColor: '$primary600',
      },
    },
    {
      bordered: true,
      variant: 'success',
      css: {
        borderColor: '$success300',
      },
    },
    {
      bordered: true,
      variant: 'danger',
      css: {
        borderColor: '$danger300',
      },
    },
    {
      bordered: true,
      variant: 'secondary',
      css: {
        borderColor: '$secondary300',
      },
    },
  ],
})

export type StyledBadgeVariants = Stitches.VariantProps<typeof StyledBadge>
export type BadgeVariantTypes = StyledBadgeVariants['variant']
