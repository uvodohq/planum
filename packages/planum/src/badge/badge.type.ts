import type * as React from 'react'

import type { CSS } from '../theme'
import type { StyledBadgeVariants } from './badge.styles'

export interface BadgeProps extends StyledBadgeVariants {
  children: React.ReactNode
  css?: CSS
}

export type BadgeVariant = StyledBadgeVariants['variant']
