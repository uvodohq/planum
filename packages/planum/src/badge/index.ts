export type { BadgeProps } from './badge'
export { default } from './badge'
