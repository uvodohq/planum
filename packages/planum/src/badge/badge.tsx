import * as React from 'react'

import type { CSS } from '../theme'
import type { StyledBadgeVariants } from './badge.styles'
import { StyledBadge } from './badge.styles'

export interface BadgeProps extends StyledBadgeVariants {
  children: React.ReactNode
  css?: CSS
}

export default function Badge(props: BadgeProps) {
  return <StyledBadge {...props} />
}
