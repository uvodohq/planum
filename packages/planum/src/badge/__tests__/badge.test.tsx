import { render, screen } from '../../test'
import { Badge } from '../badge'
import type { BadgeVariant } from '../badge.type'

// TODO: fix types, variants are not defined
const badgeVariants: BadgeVariant[] = [
  'secondary',
  'warning',
  'neutral',
  'success',
  'danger',
  'outlinePrimary',
]

describe('<Badge />', () => {
  test.each(badgeVariants)(
    'should renders %s variant badge with right text',
    (variant) => {
      const badgeText = variant as string

      const { container } = render(<Badge variant={variant}>{badgeText}</Badge>)
      const label = screen.getByText(badgeText)

      expect(label).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    },
  )
})
