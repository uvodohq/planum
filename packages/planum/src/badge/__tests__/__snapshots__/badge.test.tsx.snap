// Vitest Snapshot v1

exports[`<Badge /> > should renders danger variant badge with right text 1`] = `
<div>
  <span
    class="c-iPYhUp c-kLRYwF c-dtudLY c-dtudLY-eNlZJL-variant-danger c-dtudLY-gTzoIO-roundness-pill"
  >
    danger
  </span>
</div>
`;

exports[`<Badge /> > should renders neutral variant badge with right text 1`] = `
<div>
  <span
    class="c-iPYhUp c-kLRYwF c-dtudLY c-dtudLY-eQBtDl-variant-neutral c-dtudLY-gTzoIO-roundness-pill"
  >
    neutral
  </span>
</div>
`;

exports[`<Badge /> > should renders outlinePrimary variant badge with right text 1`] = `
<div>
  <span
    class="c-iPYhUp c-kLRYwF c-dtudLY c-dtudLY-jJsbnp-variant-outlinePrimary c-dtudLY-gTzoIO-roundness-pill"
  >
    outlinePrimary
  </span>
</div>
`;

exports[`<Badge /> > should renders secondary variant badge with right text 1`] = `
<div>
  <span
    class="c-iPYhUp c-kLRYwF c-dtudLY c-dtudLY-ebPlNg-variant-secondary c-dtudLY-gTzoIO-roundness-pill"
  >
    secondary
  </span>
</div>
`;

exports[`<Badge /> > should renders success variant badge with right text 1`] = `
<div>
  <span
    class="c-iPYhUp c-kLRYwF c-dtudLY c-dtudLY-jhsIH-variant-success c-dtudLY-gTzoIO-roundness-pill"
  >
    success
  </span>
</div>
`;

exports[`<Badge /> > should renders warning variant badge with right text 1`] = `
<div>
  <span
    class="c-iPYhUp c-kLRYwF c-dtudLY c-dtudLY-gHSaia-variant-warning c-dtudLY-gTzoIO-roundness-pill"
  >
    warning
  </span>
</div>
`;
